### PR TITLE
【autograd】add select_p、eq_p、pow_p primitive operator for new autograd

### PR DESCRIPTION
**Pull request by betterpig**
### PR types

New features
### PR changes

OPs
### Describe

- add select_p、eq_p、pow_p primitive operator for new autograd. Pow_p depends on log_p, select_p and eq_p
- use autograd to test grad with high order
- though eq has no jvp or transpose rule, it should define jvp rule and return zeros
- paddle.pow, paddle.where has no grad_grad_kernel, so can not compute forward grad use double trick
- In where op, grad of condition computed by paddle.static.gradients is None, and paddle.incubate.autograd.grad will replace None with zeros while transpose will just return None because cond_dot is unused, that is a diff. So wrap it and use fixed condition.
- set_tests_properties(test_primapi PROPERTIES TIMEOUT 60) because we will add new p_op test in test_primapi continuously and test_grad_with_high_order may cost a lot of time. Right now this test only cost aoubt 15s.

### CODE REVIEW

File: python/paddle/incubate/autograd/primops.py [hunk 1]
```diff
@@ -310,3 +310,26 @@ def scatter_add(x, y, indextensor, axis, out=None):
...
     return out
+
+
+@REGISTER_FN('select_p', 'Condition', 'X', 'Y', 'Z')
+def select(cond, x, y, out=None):
+    assert len(cond.shape) == len(x.shape), (
+        f'len(cond.shape) should be equal to len(x.shape), '
+        f'but len(cond.shape)={len(cond.shape)} and len(x.shape)={len(x.shape)}.'
+    )
+    assert len(x.shape) == len(y.shape), (
```
Comment (cxxly) [inline python/paddle/incubate/autograd/primops.py]: Do not use assert statements for validating argument values of a public API. assert is used to ensure internal correctness, not to enforce correct usage nor to indicate that some unexpected event occurred.

https://google.github.io/styleguide/pyguide.html

File: python/paddle/incubate/autograd/primrules.py [hunk 2]
```diff
@@ -285,6 +285,11 @@ def num_el(shape):
         raise RuntimeError('Only support lower l2/l1 norm currently')
 
 
+@REGISTER_ORIG2PRIM('where')
```
Comment (cxxly) [inline python/paddle/incubate/autograd/primrules.py:299]: TODO： support broadcast

File: python/paddle/fluid/tests/unittests/autograd/test_primapi.py [hunk 3]
```diff
@@ -151,7 +151,9 @@ class TestForwardGrad(unittest.TestCase):
 
     @classmethod
     def setUpClass(cls):
-        cls.xs = tuple(x.astype(cls.dtype) for x in cls.xs)
+        cls.xs = tuple(
+            x.astype(cls.dtype) if str(x.dtype) != 'bool' else x
+            for x in cls.xs)
```
Comment (cxxly) [inline python/paddle/fluid/tests/unittests/autograd/test_primapi.py]: 修改成只有在float类型需要做类型转换，不要通过反例来排除一些不需要的类型转换

File: python/paddle/fluid/tests/unittests/autograd/test_primapi.py [hunk 4]
```diff
@@ -365,6 +367,129 @@ def actual():
...
 
+# In where op, grad of condition computed by paddle.static.gradients is None,
+# and paddle.incubate.autograd.grad will replace None with zeros while transpose
+# will just return None because cond_dot is unused, that is a diff.
+@utils.place(config.DEVICES)
+@utils.parameterize(
+    (utils.TEST_CASE_NAME, 'fun', 'xs', 'v', 'dtype'),
+    (('select', paddle.where, (np.random.randn(3, 4) > 0, np.random.randn(
+        3, 4), np.random.randn(3, 4)), None, 'float32'), ))
+class TestGrad(unittest.TestCase):
```
Comment (cxxly) [inline python/paddle/fluid/tests/unittests/autograd/test_primapi.py]: TestSelectGrad

File: python/paddle/incubate/autograd/primops.py (hunk 1 above)
Comment (betterpig) [reply python/paddle/incubate/autograd/primops.py]: 已改成使用 raise ValueError

File: python/paddle/incubate/autograd/primrules.py (hunk 2 above)
Comment (betterpig) [reply python/paddle/incubate/autograd/primrules.py:299]: done

File: python/paddle/fluid/tests/unittests/autograd/test_primapi.py (hunk 3 above)
Comment (betterpig) [reply python/paddle/fluid/tests/unittests/autograd/test_primapi.py]: done

File: python/paddle/fluid/tests/unittests/autograd/test_primapi.py (hunk 4 above)
Comment (betterpig) [reply python/paddle/fluid/tests/unittests/autograd/test_primapi.py]: done

**Comment by cxxly** [comment]
修改PR标题和描述，不止一个算子

File: python/paddle/fluid/tests/unittests/autograd/test_jvp_and_transpose.py [hunk 5]
```diff
@@ -867,5 +867,90 @@ def init_data(self):
...
+            'mul_p',
+            'pow_p',
+            'mul_p',
+            'mul_p',
+            'log_p',
+            'add_p'
+            # transpose op:
+        ]
+
+
```
Comment (cxxly) [inline python/paddle/fluid/tests/unittests/autograd/test_jvp_and_transpose.py:984]: eq_p jvp测试先写上，保持测试用例完整性；后续我统一修改

File: python/paddle/fluid/tests/unittests/autograd/test_primapi.py [hunk 6]
```diff
@@ -16,6 +16,8 @@
 import unittest
 
 import numpy as np
+import autograd.numpy as agnp
```
Comment (cxxly) [inline python/paddle/fluid/tests/unittests/autograd/test_primapi.py]: 别名就叫做 np_autograd吧，agnp不是业界惯用缩写

File: python/paddle/fluid/tests/unittests/autograd/test_primapi.py [hunk 7]
```diff
@@ -148,11 +150,15 @@ def without_program_guard():
...
 ))
+# paddle.where, paddle.pow has no double grad definition,
+# can not compute forward grad use double trick
 class TestForwardGrad(unittest.TestCase):
 
     @classmethod
     def setUpClass(cls):
-        cls.xs = tuple(x.astype(cls.dtype) for x in cls.xs)
+        cls.xs = tuple(
+            x.astype(cls.dtype) if str(x.dtype) == 'float' else x
```
Comment (cxxly) [inline python/paddle/fluid/tests/unittests/autograd/test_primapi.py]: 修改成`x.dtype == float`， str(x.dtype)得到的是"float32/float64"

File: python/paddle/fluid/tests/unittests/autograd/test_primapi.py [hunk 8]
```diff
@@ -377,105 +395,201 @@ def tearDown(self):
...
+            x.astype(cls.dtype) if x.dtype != bool else x for x in cls.xs)
+        cls._rtol = config.TOLERANCE.get(str(
+            cls.dtype)).get("first_order_grad").get("rtol")
+        cls._atol = config.TOLERANCE.get(str(
+            cls.dtype)).get("first_order_grad").get("atol")
+
+    def test_grad(self):
+
+        def expected():
+            from autograd import elementwise_grad as egrad
```
Comment (cxxly) [inline python/paddle/fluid/tests/unittests/autograd/test_primapi.py]: 非必要情况，不要直接导入一个变量/函数/类，导入module，通过moudle.xx方式调用，参考goolge代码规范

File: python/paddle/incubate/autograd/primrules.py [hunk 9]
```diff
@@ -18,7 +18,7 @@
 from .primops import (add, broadcast, concat, cos, div, exp, fill_const, gather,
                       matmul, mul, neg, reduce, reshape, scatter_add, set_value,
                       sin, slice_assign, slice_select, split, sqrt, sub, tanh,
-                      transpose, log)
+                      transpose, log, select, eq, pow)
```
Comment (cxxly) [inline python/paddle/incubate/autograd/primrules.py]: pow与python内置关键子pow冲突，避免产生非预期行为，使用primops.pow方式调用

File: python/unittest_py/requirements.txt [hunk 10]
```diff
@@ -14,3 +14,4 @@ scipy>=1.5; python_version == "3.6"
 prettytable
 distro
 numpy>=1.20,<1.22; python_version >= "3.7"
+autograd
```
Comment (cxxly) [inline python/unittest_py/requirements.txt]: 加上版本号

File: python/paddle/fluid/tests/unittests/autograd/test_primapi.py [hunk 11]
```diff
@@ -367,7 +377,15 @@ def actual():
             np.testing.assert_allclose(i, j, rtol=self._rtol, atol=self._atol)
 
 
-class TestGradWithHigherOrder(unittest.TestCase):
+# In where op, grad of condition computed by paddle.static.gradients is None,
+# and paddle.incubate.autograd.grad will replace None with zeros while transpose
+# will just return None because cond_dot is unused, that is a diff.
```
Comment (cxxly) [inline python/paddle/fluid/tests/unittests/autograd/test_primapi.py]: 使用 functools.partial(func, /, *args, **keywords)[¶](https://docs.python.org/3/library/functools.html#functools.partial)屏蔽Cond参数

File: python/paddle/fluid/tests/unittests/autograd/test_primapi.py (hunk 6 above)
Comment (betterpig) [reply python/paddle/fluid/tests/unittests/autograd/test_primapi.py]: done

File: python/paddle/fluid/tests/unittests/autograd/test_primapi.py (hunk 7 above)
Comment (betterpig) [reply python/paddle/fluid/tests/unittests/autograd/test_primapi.py]: done

File: python/paddle/fluid/tests/unittests/autograd/test_primapi.py (hunk 8 above)
Comment (betterpig) [reply python/paddle/fluid/tests/unittests/autograd/test_primapi.py]: done

File: python/paddle/incubate/autograd/primrules.py (hunk 9 above)
Comment (betterpig) [reply python/paddle/incubate/autograd/primrules.py]: done

File: python/unittest_py/requirements.txt (hunk 10 above)
Comment (betterpig) [reply python/unittest_py/requirements.txt]: done

File: python/paddle/fluid/tests/unittests/autograd/test_jvp_and_transpose.py (hunk 5 above)
Comment (betterpig) [reply python/paddle/fluid/tests/unittests/autograd/test_jvp_and_transpose.py:984]: done

File: python/paddle/fluid/tests/unittests/autograd/test_primapi.py (hunk 11 above)
Comment (betterpig) [reply python/paddle/fluid/tests/unittests/autograd/test_primapi.py]: 改成了用lambda表达式，并删除了TestSelectGrad。

**Comment by betterpig** [comment]
> 修改PR标题和描述，不止一个算子

done

**Review by cxxly — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by XieYunshen — APPROVED** [review]
LGTM for `set_tests_properties(test_primapi PROPERTIES TIMEOUT 60)`

**Pull request merged**